File: 2724-convert-an-array-into-a-2d-array-with-conditions/convert-an-array-into-a-2d-array-with-conditions.py
class Solution:
    def findMatrix(self, nums: List[int]) -> List[List[int]]:
        mp = defaultdict(int)
        res = []
        for num in nums:
            row = mp[num]
            if row == len(res):
                res.append([])
            res[row].append(num)
            mp[num] += 1        
        
        return res

File: 2724-convert-an-array-into-a-2d-array-with-conditions/convert-an-array-into-a-2d-array-with-conditions.scala
import scala.annotation.tailrec
import scala.collection.mutable.ListBuffer

object Solution {
    def findMatrix(nums: Array[Int]): List[List[Int]] = {
        val mp = nums.groupBy(identity)
                     .view
                     .mapValues(_.length)
                     .toMap

        @tailrec
        def loop(mp: Map[Int, Int], res: ListBuffer[List[Int]]): ListBuffer[List[Int]] = {
            if (mp.isEmpty) res
            else {
                val row = mp.keys.toList
                val freshMp = mp.view
                                .mapValues(_ - 1)
                                .filter{case (_, freq) => freq > 0}
                                .toMap
                loop(freshMp, res += row)
            }
        }
        
        loop(mp, res = new ListBuffer[List[Int]]).toList
    }
}